README.md
# Slippy Grid

[![Build Status](https://travis-ci.org/DenisCarriere/slippy-grid.svg?branch=master)](https://travis-ci.org/DenisCarriere/slippy-grid)
[![npm version](https://badge.fury.io/js/slippy-grid.svg)](https://badge.fury.io/js/slippy-grid)
[![MIT licensed](https://img.shields.io/badge/license-MIT-blue.svg)](https://raw.githubusercontent.com/DenisCarriere/slippy-grid/master/LICENSE)

[![Standard - JavaScript Style Guide](https://cdn.rawgit.com/feross/standard/master/badge.svg)](https://github.com/feross/standard)

Create slippy tile grid iterators from BBox.

## Install

**npm**

```bash
$ npm install --save slippy-grid
```

**web browser [ES6](https://kangax.github.io/compat-table/es6)**

```html
<script src="https://unpkg.com/slippy-grid/docs/slippy-grid.min.js"></script>
```

## API

<!-- Generated by documentation.js. Update this documentation by updating the source code. -->

#### Table of Contents

-   [single](#single)
-   [geojson](#geojson)
-   [getChildren](#getchildren)
-   [all](#all)
-   [bulk](#bulk)
-   [levels](#levels)
-   [count](#count)

### single

Creates an Iterator of Tiles from a given BBox

**Parameters**

-   `extent` **(BBox | [Array](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array)&lt;BBox> | GeoJSON)** BBox [west, south, east, north] order or GeoJSON Polygon
-   `minZoom` **[number](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number)** Minimum Zoom
-   `maxZoom` **[number](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number)** Maximum Zoom

**Examples**

```javascript
const grid = slippyGrid.single([-180.0, -90.0, 180, 90], 3, 8)
const {value, done} = grid.next()
//=value [x, y, z]
//=done true/false
```

Returns **Iterator&lt;Tile>** Iterable Grid of Tiles from extent

### geojson

Creates an Iterator of Tiles from a given GeoJSON

**Parameters**

-   `extent` **FeatureCollection&lt;(Polygon | MultiPolygon)>** GeoJSON Polygon(s)
-   `minZoom` **[number](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number)** Minimum Zoom
-   `maxZoom` **[number](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number)** Maximum Zoom

**Examples**

```javascript
const grid = slippyGrid.geojson(poly, 3, 8)
const {value, done} = grid.next()
// value => [x, y, z]
// done => true/false
```

Returns **Iterator&lt;Tile>** Iterable Grid of Tiles from GeoJSON

### getChildren

Creates an Iterator of children Tiles from a given parent tile

**Parameters**

-   `parentTile` **GeoJSON** Tile
-   `maxZoom` **[number](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number)** Maximum Zoom

**Examples**

```javascript
const parentTile = [0, 572, 10]
const maxZoom = 12
const grid = slippyGrid.getChildren(parentTile, maxZoom)
const {value, done} = grid.next()
// value => [x, y, z]
// done => true/false
```

Returns **Iterator&lt;Tile>** Iterable of Tiles

### all

All Tiles from a given BBox

**Parameters**

-   `extent` **(BBox | [Array](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array)&lt;BBox> | GeoJSON)** BBox [west, south, east, north] order or GeoJSON Polygon
-   `minZoom` **[number](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number)** Minimum Zoom
-   `maxZoom` **[number](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number)** Maximum Zoom

**Examples**

```javascript
const tiles = slippyGrid.all([-180.0, -90.0, 180, 90], 3, 8)
//=tiles
```

Returns **[Array](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array)&lt;Tile>** Tiles from extent

### bulk

Creates a bulk Iterator of Tiles from a given BBox

**Parameters**

-   `extent` **(BBox | [Array](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array)&lt;BBox> | GeoJSON)** BBox [west, south, east, north] order or GeoJSON Polygon
-   `minZoom` **[number](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number)** Minimum Zoom
-   `maxZoom` **[number](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number)** Maximum Zoom
-   `size` **[number](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number)** Maximum size for bulk Tiles

**Examples**

```javascript
const grid = slippyGrid.bulk([-180.0, -90.0, 180, 90], 3, 8, 5000)
const {value, done} = grid.next()
//=value Array<[x, y, z]>
//=done true/false
```

Returns **Iterator&lt;[Array](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array)&lt;Tile>>** Bulk Iterable Grid of Tiles from extent

### levels

Creates a grid level pattern of arrays

**Parameters**

-   `extent` **(BBox | [Array](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array)&lt;BBox> | GeoJSON)** BBox [west, south, east, north] order or GeoJSON Polygon
-   `minZoom` **[number](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number)** Minimum Zoom
-   `maxZoom` **[number](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number)** Maximum Zoom

**Examples**

```javascript
const levels = slippyGrid.levels([-180.0, -90.0, 180, 90], 3, 8)
//=levels
```

Returns **[Array](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array)&lt;GridLevel>** Grid Level

### count

Counts the total amount of tiles from a given BBox

**Parameters**

-   `extent` **(BBox | [Array](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array)&lt;BBox> | GeoJSON)** BBox [west, south, east, north] order or GeoJSON Polygon
-   `minZoom` **[number](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number)** Minimum Zoom
-   `maxZoom` **[number](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number)** Maximum Zoom
-   `quick` **[number](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number)** Enable quick count if greater than number (optional, default `1000`)

**Examples**

```javascript
const count = slippyGrid.count([-180.0, -90.0, 180, 90], 3, 8)
//=count 563136
```

Returns **[number](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number)** Total tiles from BBox


CHANGELOG.md

# Changelog

## 2.2.0 - 2017-10-23

- Fix Fiji GeoJSON extent using optimal tile cover
- Support GeoJSON maximum/optimal tile cover

## 2.1.0 - 2017-03-30

- Drop unique tiles (caused memory leaks on 1.3M+)

## 2.0.0 - 2017-03-17

- Add `all()` method to return all tiles in a simple `Array`
- Support Fiji Extent & any extent that share +180 & -180 degrees
- Dropped GeoJSON precise extent (only support BBox of a GeoJSON)

## 1.4.0 - 2017-03-10

- Fix `levels()` & `count()` not returning the correct results.

## 1.3.0 - 2017-03-04

- Output browser bundle to docs UMD
- Change rollup config
- Rebuild library to pure ES6
- Improved documentation
- Add test case for slippyGrid.levels() => World Zoom Level 1

## 1.2.1 - 2017-03-02

- Removed `fastCount` from Typescript definition
- Lowered quick count to 1000

## 1.2.0 - 2017-02-22

- Support overlapping GeoJSON properties with multiple different zoom levels
- Filter by minZoom & maxZoom via GeoJSON properties

## 1.1.0 - 2017-02-21

- Support Multiple GeoJSON Polygon
- Added GeoJSON input support

## 1.0.0 - 2017-02-18

- Stable release
- Initialize library

index.js
const cover = require('@mapbox/tile-cover')
const tilebelt = require('@mapbox/tilebelt')
const turfBBox = require('@turf/bbox')
const { featureEach } = require('@turf/meta')
const { range, lngLatToTile, googleToTile } = require('global-mercator')

function wrapTile (tile) {
  let [x, y, z] = tile
  const z2 = Math.pow(2, z)
  x = x % z2
  if (x < 0) x = x + z2
  return [x, y, z]
}

/**
 * Creates an Iterator of Tiles from a given BBox
 *
 * @param {BBox|BBox[]|GeoJSON} extent BBox [west, south, east, north] order or GeoJSON Polygon
 * @param {number} minZoom Minimum Zoom
 * @param {number} maxZoom Maximum Zoom
 * @returns {Iterator<Tile>} Iterable Grid of Tiles from extent
 * @example
 * const grid = slippyGrid.single([-180.0, -90.0, 180, 90], 3, 8)
 * const {value, done} = grid.next()
 * //=value [x, y, z]
 * //=done true/false
 */
function * single (extent, minZoom, maxZoom) {
  if (Array.isArray(extent)) {
    for (const [columns, rows, zoom] of levels(extent, minZoom, maxZoom)) {
      for (const row of rows) {
        for (const column of columns) {
          const tile = [column, row, zoom]
          yield tile
        }
      }
    }
  } else if (extent.type) {
    yield * geojson(extent, minZoom, maxZoom)
  } else {
    throw new Error('unknown extent')
  }
}

/**
 * Creates an Iterator of Tiles from a given GeoJSON
 *
 * @param {FeatureCollection<Polygon|MultiPolygon>} extent GeoJSON Polygon(s)
 * @param {number} minZoom Minimum Zoom
 * @param {number} maxZoom Maximum Zoom
 * @returns {Iterator<Tile>} Iterable Grid of Tiles from GeoJSON
 * @example
 * const grid = slippyGrid.geojson(poly, 3, 8)
 * const {value, done} = grid.next()
 * // value => [x, y, z]
 * // done => true/false
 */
function * geojson (extent, minZoom, maxZoom) {
  const features = (extent.type === 'FeatureCollection') ? extent.features : [extent]
  for (const feature of features) {
    const parentTiles = cover.tiles(feature.geometry, {
      min_zoom: minZoom,
      max_zoom: maxZoom
    })
    for (const parentTile of parentTiles) {
      const zoom = parentTile[2]
      if (zoom > maxZoom) continue
      yield wrapTile(googleToTile(parentTile))
      yield * getChildren(parentTile, maxZoom)
    }
  }
}

/**
 * Creates an Iterator of children Tiles from a given parent tile
 *
 * @param {GeoJSON} parentTile Tile
 * @param {number} maxZoom Maximum Zoom
 * @returns {Iterator<Tile>} Iterable of Tiles
 * @example
 * const parentTile = [0, 572, 10]
 * const maxZoom = 12
 * const grid = slippyGrid.getChildren(parentTile, maxZoom)
 * const {value, done} = grid.next()
 * // value => [x, y, z]
 * // done => true/false
 */
function * getChildren (parentTile, maxZoom) {
  for (const tile of tilebelt.getChildren(parentTile)) {
    const zoom = tile[2]
    if (zoom > maxZoom) continue
    if (zoom < maxZoom) yield * getChildren(tile, maxZoom)
    yield wrapTile(googleToTile(tile))
  }
}

/**
 * All Tiles from a given BBox
 *
 * @param {BBox|BBox[]|GeoJSON} extent BBox [west, south, east, north] order or GeoJSON Polygon
 * @param {number} minZoom Minimum Zoom
 * @param {number} maxZoom Maximum Zoom
 * @returns {Array<Tile>} Tiles from extent
 * @example
 * const tiles = slippyGrid.all([-180.0, -90.0, 180, 90], 3, 8)
 * //=tiles
 */
function all (extent, minZoom, maxZoom) {
  const tiles = []
  const grid = single(extent, minZoom, maxZoom)
  while (true) {
    const {value, done} = grid.next()
    if (done) break
    tiles.push(value)
  }
  return tiles
}

/**
 * Creates a bulk Iterator of Tiles from a given BBox
 *
 * @param {BBox|BBox[]|GeoJSON} extent BBox [west, south, east, north] order or GeoJSON Polygon
 * @param {number} minZoom Minimum Zoom
 * @param {number} maxZoom Maximum Zoom
 * @param {number} size Maximum size for bulk Tiles
 * @returns {Iterator<Tile[]>} Bulk Iterable Grid of Tiles from extent
 * @example
 * const grid = slippyGrid.bulk([-180.0, -90.0, 180, 90], 3, 8, 5000)
 * const {value, done} = grid.next()
 * //=value Array<[x, y, z]>
 * //=done true/false
 */
function * bulk (extent, minZoom, maxZoom, size) {
  const iterable = single(extent, minZoom, maxZoom)
  let container = []
  let i = 0
  while (true) {
    i++
    const { value, done } = iterable.next()
    if (value) {
      container.push(value)
    }
    if (i % size === 0) {
      yield container
      container = []
    }
    if (done) {
      yield container
      break
    }
  }
}

/**
 * Creates a grid level pattern of arrays
 *
 * @param {BBox|BBox[]|GeoJSON} extent BBox [west, south, east, north] order or GeoJSON Polygon
 * @param {number} minZoom Minimum Zoom
 * @param {number} maxZoom Maximum Zoom
 * @returns {GridLevel[]} Grid Level
 * @example
 * const levels = slippyGrid.levels([-180.0, -90.0, 180, 90], 3, 8)
 * //=levels
 */
function levels (extent, minZoom, maxZoom) {
  const extents = []
  if (extent === undefined) throw new Error('extent is required')
  if (minZoom === undefined) throw new Error('minZoom is required')
  if (maxZoom === undefined) throw new Error('maxZoom is required')

  // Single Array
  if (extent.length === 4 && extent[0][0] === undefined) { extents.push({bbox: extent, minZoom, maxZoom}) }

  // Multiple Array
  if (extent.length && extent[0][0] !== undefined) { extent.map(inner => extents.push({bbox: inner, minZoom, maxZoom})) }

  // GeoJSON
  featureEach(extent, feature => {
    const bbox = turfBBox(feature)
    const featureMinZoom = feature.properties.minZoom || feature.properties.minzoom || minZoom
    const featureMaxZoom = feature.properties.maxZoom || feature.properties.maxzoom || maxZoom
    extents.push({bbox, minZoom: featureMinZoom, maxZoom: featureMaxZoom})
  })
  const levels = []
  for (const {bbox, minZoom, maxZoom} of extents) {
    let [x1, y1, x2, y2] = bbox
    for (const zoom of range(minZoom, maxZoom + 1)) {
      if (y2 > 85) y2 = 85
      if (y2 < -85) y2 = -85
      const t1 = lngLatToTile([x1, y1], zoom)
      const t2 = lngLatToTile([x2, y2], zoom)

      // Columns
      let columns = []
      // Fiji - World divided into two parts
      if (t1[0] > t2[0]) {
        // right world +180 degrees
        const maxtx = Math.pow(2, zoom) - 1
        const rightColumns = range(t1[0], maxtx + 1)
        rightColumns.forEach(column => columns.push(column))

        // left world -180 degrees
        const mintx = 0
        const leftColumns = range(mintx, t2[0] + 1)
        leftColumns.forEach(column => columns.push(column))
      } else {
        // Normal World
        const mintx = Math.min(t1[0], t2[0])
        const maxtx = Math.max(t1[0], t2[0])
        columns = range(mintx, maxtx + 1)
      }

      // Rows
      const minty = Math.min(t1[1], t2[1])
      const maxty = Math.max(t1[1], t2[1])
      const rows = range(minty, maxty + 1)

      levels.push([columns, rows, zoom])
    }
  }
  return levels
}

/**
 * Counts the total amount of tiles from a given BBox
 *
 * @param {BBox|BBox[]|GeoJSON} extent BBox [west, south, east, north] order or GeoJSON Polygon
 * @param {number} minZoom Minimum Zoom
 * @param {number} maxZoom Maximum Zoom
 * @param {number} [quick=1000] Enable quick count if greater than number
 * @returns {number} Total tiles from BBox
 * @example
 * const count = slippyGrid.count([-180.0, -90.0, 180, 90], 3, 8)
 * //=count 563136
 */
function count (extent, minZoom, maxZoom, quick) {
  quick = quick || 1000
  let count = 0

  // Quick count
  if (quick !== -1) {
    for (const [columns, rows] of levels(extent, minZoom, maxZoom)) {
      count += rows.length * columns.length
    }
    if (count > quick) { return count }
  }

  // Accurate count
  count = 0
  const grid = single(extent, minZoom, maxZoom)
  while (true) {
    const {done} = grid.next()
    if (done) { break }
    count++
  }
  return count
}

module.exports = {single, bulk, levels, count, all, geojson, getChildren}
